src/NHibernate/Multi/QueryBatch.cs

```diff
@@ -161,11 +166,7 @@ protected void DoExecute()
 					{
 						foreach (var multiSource in _queries)
 						{
-							foreach (var resultSetHandler in multiSource.GetResultSetHandler())
-							{
-								rowCount += resultSetHandler(reader);
-								reader.NextResult();
```

Comment:
I was trying to avoid code duplication in `IQueryBatchItem`  implementations  and from this point -  `reader.NextResult` call should be responsibility of `IQueryBatch' implemention. And I've made [suggestion](https://github.com/nhibernate/nhibernate-core/pull/1718#discussion_r201482522) that should allow to keep old interface/behavior intact.
But it seems your refactoring fixed one more issue - no async counterpart was generated for old `GetResultSetHandler` method.
So this implementation might be a better option as it's suppored by async generator.

Reply:
Well, about `NextResult`, I understand it could be considered the responsibility of the batcher. But at the same time, that is the queries who know how many result sets they have to consume. So there is no clear cut on that subject from my point of view.